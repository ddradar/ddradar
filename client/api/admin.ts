import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
