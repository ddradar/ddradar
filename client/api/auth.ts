import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

type AuthResult = {
  clientPrincipal: Api.ClientPrincipal | null
}

/**
 * Call "Get Credentials" API.
 * @see https://docs.microsoft.com/azure/static-web-apps/user-information
 */
export async function getClientPrincipal(
  $http: Pick<NuxtHTTPInstance, '$get'>
) {
  const { clientPrincipal } = await $http.$get<AuthResult>('/.auth/me')
  return clientPrincipal
}
