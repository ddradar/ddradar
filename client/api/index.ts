export const apiPrefix = '/api/v1'
