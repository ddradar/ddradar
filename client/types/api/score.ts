/**
 * Object type returned by `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/getChartScore/README.md
 */
export type UserScore = {
  userId: string
  userName: string
  /** Song id that depend on official site. ^([01689bdiloqDIOPQ]*){32}$ */
  songId: string
  songName: string
  playStyle: 1 | 2
  difficulty: 0 | 1 | 2 | 3 | 4
  /** Normal score */
  score: number
  exScore?: number
  maxCombo?: number
  /** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
  clearLamp: ClearLamp
  /** Clear rank (E-AAA) */
  rank: string
}

/** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
export type ClearLamp = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7

/**
 * Request body to `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/postChartScore/README.md
 */
export type Score = Pick<
  UserScore,
  'score' | 'exScore' | 'maxCombo' | 'clearLamp' | 'rank'
>

export function getDanceLevel(score: number) {
  const rankList = [
    { border: 990000, rank: 'AAA' },
    { border: 950000, rank: 'AA+' },
    { border: 900000, rank: 'AA' },
    { border: 890000, rank: 'AA-' },
    { border: 850000, rank: 'A+' },
    { border: 800000, rank: 'A' },
    { border: 790000, rank: 'A-' },
    { border: 750000, rank: 'B+' },
    { border: 700000, rank: 'B' },
    { border: 690000, rank: 'B-' },
    { border: 650000, rank: 'C+' },
    { border: 600000, rank: 'C' },
    { border: 590000, rank: 'C-' },
    { border: 550000, rank: 'D+' },
  ] as const
  for (const { border, rank } of rankList) {
    if (score >= border) return rank
  }
  return 'D'
}
