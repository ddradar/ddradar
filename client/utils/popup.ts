import type { BuefyNamespace } from 'buefy/types'

const option = { position: 'is-top', hasIcon: true } as const

export function success(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-success', ...option })
}

export function warning(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-warning', ...option })
}

export function danger(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-danger', ...option })
}
