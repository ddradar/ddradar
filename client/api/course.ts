import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function getCourseType(type: number) {
  return type === 1 ? 'NONSTOP' : type === 2 ? '段位認定' : ''
}

/**
 * Call "Get Course List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses/
 */
export function getCourseList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series?: 16 | 17,
  type?: 1 | 2
) {
  const searchParams = new URLSearchParams()
  if (series) searchParams.append('series', `${series}`)
  if (type) searchParams.append('type', `${type}`)
  return $http.$get<Api.CourseListData[]>(`${apiPrefix}/courses`, {
    searchParams,
  })
}

/**
 * Call "Get Course Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses__id/
 */
export function getCourseInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.CourseInfo>(`${apiPrefix}/courses/${id}`)
}
