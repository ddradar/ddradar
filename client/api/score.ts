import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--get/
 */
export function getChartScore(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  scope?: 'private' | 'medium' | 'full'
) {
  const query = scope ? `?scope=${scope}` : ''
  return $http.$get<Api.ScoreInfo[]>(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}${query}`
  )
}

/**
 * Call "Post Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--post/
 */
export function postChartScore(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  score: Api.ScoreBody
) {
  return $http.$post(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}`,
    score
  )
}

/**
 * Call "Delete Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--delete/
 */
export function deleteChartScore(
  $http: Pick<NuxtHTTPInstance, 'delete'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4
) {
  return $http.delete(`/api/v1/scores/${songId}/${playStyle}/${difficulty}`)
}

/**
 * Call "Post Song Scores" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id--post/
 */
export function postSongScores(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  scores: Api.ScoreListBody[]
) {
  return $http.$post<Api.ScoreInfo[]>(`${apiPrefix}/scores/${songId}`, scores)
}
