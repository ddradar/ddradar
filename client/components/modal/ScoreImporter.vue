<template>
  <div class="modal-card">
    <header class="modal-card-head">
      <h1 class="modal-card-title">{{ $t('title') }}</h1>
    </header>
    <section class="modal-card-body">
      <div class="content">
        <ol>
          <li>{{ $t('process_1') }}</li>
          <i18n path="process_2" tag="li">
            <a target="_blank" :href="musicDetailUri">
              {{ $t('process_2_0') }}
            </a>
          </i18n>
          <li>{{ $t('process_3') }}</li>
        </ol>
      </div>
      <b-field>
        <b-input v-model="sourceCode" type="textarea" required />
      </b-field>
    </section>
    <footer class="modal-card-foot">
      <b-button
        :disabled="!sourceCode"
        :loading="loading"
        type="is-success"
        @click="importScore()"
      >
        {{ $t('regist') }}
      </b-button>
      <b-button type="is-warning" @click="close()">
        {{ $t('close') }}
      </b-button>
    </footer>
  </div>
</template>

<i18n>
{
  "ja": {
    "title": "スコアのインポート",
    "process_1": "e-amusement GATEにログインしてください。",
    "process_2": "{0}のソースコードを取得し、コピーしてください。",
    "process_2_0": "楽曲データ詳細",
    "process_3": "下のテキストボックスに、コピーしたソースコードを貼り付けてください。",
    "regist": "登録",
    "success": "インポートしました",
    "close": "閉じる"
  },
  "en": {
    "title": "Import Score",
    "process_1": "Login to e-amusement GATE.",
    "process_2": "Get {0} HTML source, and copy it.",
    "process_2_0": "Music Detail's",
    "process_3": "Paste copied text below.",
    "regist": "Import",
    "success": "Imported",
    "close": "Close"
  }
}
</i18n>

<script lang="ts">
import { Gate } from '@ddradar/core'
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import { postSongScores } from '~/api/score'
import * as popup from '~/utils/popup'

@Component
export default class ScoreImporterComponent extends Vue {
  @Prop({ required: true, type: String })
  readonly songId!: string

  @Prop({ required: true, type: Number })
  readonly playStyle!: 1 | 2

  @Prop({ required: true, type: Number })
  readonly difficulty!: 0 | 1 | 2 | 3 | 4

  @Prop({ required: false, type: Boolean, default: false })
  readonly isCourse!: boolean

  sourceCode: string | null = null
  loading: boolean = false

  get musicDetailUri() {
    /** 0 - 9 */
    const diff = (this.playStyle - 1) * 4 + this.difficulty
    return `https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/${
      this.isCourse ? 'course' : 'music'
    }_detail.html?index=${this.songId}&diff=${diff}`
  }

  async importScore() {
    if (!this.sourceCode) return

    this.loading = true

    // Convert sourceCode to Score
    let score: ReturnType<typeof Gate.musicDetailToScore>
    try {
      score = Gate.musicDetailToScore(this.sourceCode)
    } catch (error) {
      popup.warning(this.$buefy, error.message ?? error)
      this.loading = false
      return
    }

    try {
      await postSongScores(this.$http, this.songId, [score])
      popup.success(this.$buefy, this.$t('success') as string)
    } catch (error) {
      popup.danger(this.$buefy, error.message ?? error)
      this.loading = false
      return
    }

    this.loading = false
    this.close()
  }

  close() {
    // @ts-ignore
    this.$parent.close()
  }
}
</script>
