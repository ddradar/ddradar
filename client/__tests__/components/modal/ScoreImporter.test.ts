import { Gate } from '@ddradar/core'
import { createLocalVue, mount, shallowMount, Wrapper } from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import { postSongScores } from '~/api/score'
import ScoreImporter from '~/components/modal/ScoreImporter.vue'
import * as popup from '~/utils/popup'

jest.mock('@ddradar/core')
jest.mock('~/api/score')
jest.mock('~/utils/popup')
const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('/components/modal/ScoreImporter.vue', () => {
  const songId = '8Il6980di8P89lil1PDIqqIbiq1QO8lQ'
  const propsData = { songId, playStyle: 1, difficulty: 1, isCourse: false }
  const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })

  describe('snapshot test', () => {
    test.each(['ja', 'en'])(
      '{ locale: %s, sourceCode: "foo" } renders normal button',
      locale => {
        const wrapper = mount(ScoreImporter, {
          localVue,
          propsData,
          data: () => ({ sourceCode: 'foo', loading: false }),
          i18n: new VueI18n({ locale, silentFallbackWarn: true }),
        })
        expect(wrapper).toMatchSnapshot()
      }
    )
    test('{ sourceCode: null } renders disabled button', () => {
      const wrapper = mount(ScoreImporter, {
        localVue,
        propsData,
        data: () => ({ sourceCode: null, loading: false }),
        i18n,
      })
      expect(wrapper).toMatchSnapshot()
    })
    test('{ loading: true } renders loading button', () => {
      const wrapper = mount(ScoreImporter, {
        localVue,
        propsData,
        data: () => ({ sourceCode: 'foo', loading: true }),
        i18n,
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Computed
  describe('get musicDetailUri()', () => {
    test.each([
      [1, 0, false, /^.+\/music_detail\.html\?index=.{32}&diff=0$/],
      [1, 1, false, /^.+\/music_detail\.html\?index=.{32}&diff=1$/],
      [1, 2, false, /^.+\/music_detail\.html\?index=.{32}&diff=2$/],
      [1, 3, false, /^.+\/music_detail\.html\?index=.{32}&diff=3$/],
      [1, 4, false, /^.+\/music_detail\.html\?index=.{32}&diff=4$/],
      [2, 1, false, /^.+\/music_detail\.html\?index=.{32}&diff=5$/],
      [2, 2, false, /^.+\/music_detail\.html\?index=.{32}&diff=6$/],
      [2, 3, false, /^.+\/music_detail\.html\?index=.{32}&diff=7$/],
      [2, 4, false, /^.+\/music_detail\.html\?index=.{32}&diff=8$/],
      [1, 0, true, /^.+\/course_detail\.html\?index=.{32}&diff=0$/],
      [1, 1, true, /^.+\/course_detail\.html\?index=.{32}&diff=1$/],
      [1, 2, true, /^.+\/course_detail\.html\?index=.{32}&diff=2$/],
      [1, 3, true, /^.+\/course_detail\.html\?index=.{32}&diff=3$/],
      [1, 4, true, /^.+\/course_detail\.html\?index=.{32}&diff=4$/],
      [2, 1, true, /^.+\/course_detail\.html\?index=.{32}&diff=5$/],
      [2, 2, true, /^.+\/course_detail\.html\?index=.{32}&diff=6$/],
      [2, 3, true, /^.+\/course_detail\.html\?index=.{32}&diff=7$/],
      [2, 4, true, /^.+\/course_detail\.html\?index=.{32}&diff=8$/],
    ])(
      '{ playStyle: %i, difficulty: %i, isCourse: %p } returns "%s"',
      (playStyle, difficulty, isCourse, pattern) => {
        // Arrange
        const wrapper = shallowMount(ScoreImporter, {
          localVue,
          propsData: { songId, playStyle, difficulty, isCourse },
          i18n,
        })

        // Act
        // @ts-ignore
        const uri = wrapper.vm.musicDetailUri

        // Assert
        expect(uri).toMatch(pattern)
      }
    )
  })

  // Method
  describe('importScore()', () => {
    const score: ReturnType<typeof Gate.musicDetailToScore> = {
      songId: 'ld6P1lbb0bPO9doqbbPOoPb8qoDo8id0',
      playStyle: 1,
      difficulty: 0,
      score: 1000000,
      rank: 'AAA',
      clearLamp: 7,
      maxCombo: 116,
      topScore: 1000000,
    }
    let wrapper: Wrapper<ScoreImporter>
    const $http = { $post: jest.fn() }
    const $buefy = { notification: {} }
    const postMock = mocked(postSongScores)
    beforeEach(() => {
      postMock.mockClear()
      wrapper = shallowMount(ScoreImporter, {
        localVue,
        mocks: { $http, $buefy },
        propsData,
        i18n,
      })
    })

    test('does not call "Post Song Scores" API', async () => {
      // Arrange - Act
      // @ts-ignore
      await wrapper.vm.importScore()

      // Assert
      expect(mocked(Gate.musicDetailToScore)).not.toBeCalled()
      expect(postMock).not.toBeCalled()
    })
    test('calls "Post Song Scores" API', async () => {
      // Arrange
      wrapper.setData({ sourceCode: '<html></html>', loading: false })
      await wrapper.vm.$nextTick()
      const convertMock = mocked(Gate.musicDetailToScore)
      convertMock.mockReturnValue(score)
      // @ts-ignore
      wrapper.vm.$parent.close = jest.fn()

      // Act
      // @ts-ignore
      await wrapper.vm.importScore()

      // Assert
      expect(postMock).lastCalledWith($http, songId, [score])
    })
    test('shows warning message if sourceCode is invalid', async () => {
      // Arrange
      wrapper.setData({ sourceCode: '<html></html>', loading: false })
      await wrapper.vm.$nextTick()
      const warningMock = mocked(popup.warning)
      const convertMock = mocked(Gate.musicDetailToScore)
      convertMock.mockImplementation(() => {
        throw new Error('invalid')
      })

      // Act
      // @ts-ignore
      await wrapper.vm.importScore()

      // Assert
      expect(postMock).not.toBeCalled()
      expect(warningMock).lastCalledWith($buefy, 'invalid')
    })
    test('shows error message if API returns ErrorCode', async () => {
      // Arrange
      const errorMessage = '500 Server Error'
      wrapper.setData({ sourceCode: '<html></html>', loading: false })
      await wrapper.vm.$nextTick()
      const dangerMock = mocked(popup.danger)
      const convertMock = mocked(Gate.musicDetailToScore)
      convertMock.mockReturnValue(score)
      postMock.mockRejectedValue(new Error(errorMessage))

      // Act
      // @ts-ignore
      await wrapper.vm.importScore()

      // Assert
      expect(dangerMock).lastCalledWith($buefy, errorMessage)
    })
  })
})
