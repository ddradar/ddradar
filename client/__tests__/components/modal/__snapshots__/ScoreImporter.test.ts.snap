// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/modal/ScoreImporter.vue snapshot test { loading: true } renders loading button 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">スコアのインポート</h1>
  </header>
  <section class="modal-card-body">
    <div class="content">
      <ol>
        <li>e-amusement GATEにログインしてください。</li>
        <li><a target="_blank" href="https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_detail.html?index=8Il6980di8P89lil1PDIqqIbiq1QO8lQ&amp;diff=1">
            楽曲データ詳細
          </a>のソースコードを取得し、コピーしてください。</li>
        <li>下のテキストボックスに、コピーしたソースコードを貼り付けてください。</li>
      </ol>
    </div>
    <div class="field">
      <!---->
      <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" class="button is-success is-loading">
      <!----> <span>
      登録
    </span>
      <!----></button> <button type="button" class="button is-warning">
      <!----> <span>
      閉じる
    </span>
      <!----></button></footer>
</div>
`;

exports[`/components/modal/ScoreImporter.vue snapshot test { sourceCode: "foo" } renders English id locale is en 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">Import Score</h1>
  </header>
  <section class="modal-card-body">
    <div class="content">
      <ol>
        <li>Login to e-amusement GATE.</li>
        <li>Get <a target="_blank" href="https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_detail.html?index=8Il6980di8P89lil1PDIqqIbiq1QO8lQ&amp;diff=1">
            Music Detail's
          </a> HTML source, and copy it.</li>
        <li>Paste copied text below.</li>
      </ol>
    </div>
    <div class="field">
      <!---->
      <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" class="button is-success">
      <!----> <span>
      Import
    </span>
      <!----></button> <button type="button" class="button is-warning">
      <!----> <span>
      Close
    </span>
      <!----></button></footer>
</div>
`;

exports[`/components/modal/ScoreImporter.vue snapshot test { sourceCode: "foo" } renders normal button 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">スコアのインポート</h1>
  </header>
  <section class="modal-card-body">
    <div class="content">
      <ol>
        <li>e-amusement GATEにログインしてください。</li>
        <li><a target="_blank" href="https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_detail.html?index=8Il6980di8P89lil1PDIqqIbiq1QO8lQ&amp;diff=1">
            楽曲データ詳細
          </a>のソースコードを取得し、コピーしてください。</li>
        <li>下のテキストボックスに、コピーしたソースコードを貼り付けてください。</li>
      </ol>
    </div>
    <div class="field">
      <!---->
      <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" class="button is-success">
      <!----> <span>
      登録
    </span>
      <!----></button> <button type="button" class="button is-warning">
      <!----> <span>
      閉じる
    </span>
      <!----></button></footer>
</div>
`;

exports[`/components/modal/ScoreImporter.vue snapshot test { sourceCode: null } renders disabled button 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">スコアのインポート</h1>
  </header>
  <section class="modal-card-body">
    <div class="content">
      <ol>
        <li>e-amusement GATEにログインしてください。</li>
        <li><a target="_blank" href="https://p.eagate.573.jp/game/ddr/ddra20/p/playdata/music_detail.html?index=8Il6980di8P89lil1PDIqqIbiq1QO8lQ&amp;diff=1">
            楽曲データ詳細
          </a>のソースコードを取得し、コピーしてください。</li>
        <li>下のテキストボックスに、コピーしたソースコードを貼り付けてください。</li>
      </ol>
    </div>
    <div class="field">
      <!---->
      <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" disabled="disabled" class="button is-success">
      <!----> <span>
      登録
    </span>
      <!----></button> <button type="button" class="button is-warning">
      <!----> <span>
      閉じる
    </span>
      <!----></button></footer>
</div>
`;
